core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -104,10 +110,14 @@
   // TODO(gpang): remove this set, and just use the indexes.
   private final ConcurrentHashSet<T> mObjects = new ConcurrentHashSet<>(8, 0.95f, 8);
   /**
-   * Map from {@link FieldIndex} to the index. An index is a map from index value to set of
-   * objects with that index value.
+   * Map from {@link FieldIndex} to the index. An index is a map from index value to one or a
+   * set of objects with that index value. A Unique index is an index where each index value
+   * only maps to one object. A NonUnique index is an index where a index value can map to one or
+   * more objects.
    */
-  private final Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> mIndexMap;
+  private final Map<FieldIndex<T>, ConcurrentHashMap<Object, T>> mIndexMapUnique;
+  private final Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>>
+      mIndexMapNonUnique;
 
   /**
    * An interface representing an index for this {@link IndexedSet}, each index for this set must
```

Comment:
This is not as elegant as the current variadic list of indices, but gives us type safety, which I think is more important to prevent creating invalid IndexedSets.